src/common/buffer.cc

```diff
@@ -404,185 +471,539 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
+        }
+        if (r > 0) {
+          pipe_read_elements.back().size += r;
+          total_moved += r;
+          dst_pos += r;
+          actual_content_size += r;
+          if (src_offset != -1)
+            src_offset += r;
+          count -= r;
+        } else if (r == 0) {
+          //something got closed...
+          break;
+        } else {
+          //on linux, if you try to splice to pipe, you can get EAGAIN
+          //if you started from offset that was unaligned, last page cannot be filled
+          if (is_pipe_full(pipe_curr_write_fd)) {
+            //it is just pipe blocked
+            lclose(pipe_curr_write_fd);
+            pipe_curr_write_fd = -1;
+          } else {
+            int e = -errno;
+            if (e == -EAGAIN) {
+              if (total_moved == 0)
+                return e;
+              else
+                break;
+            } else {
+              lclose(pipe_curr_write_fd);
+              pipe_curr_write_fd = -1;
+              get_data();
+              if (total_moved!=0)
+                return total_moved;
+              return buffer::raw::insert_from_fd(dst_pos, count, fd, src_offset);
+            }
+          }
+        }
+      } else {
+        if (total_moved>0)
+          return total_moved;
+
+        //nothing moved; have to resort to read/write
+        char buffer[CEPH_PAGE_SIZE*2];
+        ssize_t r, w;
+        if (data_avail < to_move)
+          to_move = data_avail;
+        //to_move is always <= 2*CEPH_PAGE_SIZE
+        if (src_offset == -1) {
+          r = read(fd, &buffer, to_move);
+        } else {
+          r = pread(fd, &buffer, to_move, src_offset);
+        }
+        if (r > 0) {
+          total_moved += r;
+          dst_pos += r;
+          actual_content_size += r;
+          if (src_offset != -1)
+            src_offset += r;
+          count -= r;
+          w = write(pipe_curr_write_fd, &buffer, r);
+          if (w == r) {
+            pipe_read_elements.back().size += r;
+          } else {
+            if (w == -1) {
+              w = 0;
+            }
+            lclose(pipe_curr_write_fd);
+            append_pipe();
+            ssize_t wf = write(pipe_curr_write_fd, &buffer + w, r - w);
+            pipe_read_elements.back().size += wf;
+            //we expect that in NEW pipe there will be enough space for CEPH_PAGE_SIZE*2
+            assert(wf == r - w);
+          }
+        } else if (r==0) {
+          return 0;
+        } else {
+          return -errno;
+        }
+      }
     }
+   return total_moved;
+  }
 
-    void close_pipe(int *fds) {
-      if (fds[0] >= 0)
-	VOID_TEMP_FAILURE_RETRY(::close(fds[0]));
-      if (fds[1] >= 0)
-	VOID_TEMP_FAILURE_RETRY(::close(fds[1]));
-    }
-    char *copy_pipe(int *fds) {
-      /* preserve original pipe contents by copying into a temporary
-       * pipe before reading.
-       */
-      int tmpfd[2];
-      int r;
+  buffer::raw* clone_empty() {
```

Comment:
When `CEPH_HAVE_SPLICE` is enabled, `::clone_empty()` returns `NULL`. raw::clone() does not check the return value of `clone_empty()` for this case. My need to add the check, such as:

```Cpp
    raw *clone() {
      raw *c = clone_empty();
#ifdef CEPH_HAVE_SPLICE
      if (!c) return NULL;
#endif
      memcpy(c->data, data, len);
      return c;
    }
```